src/hotspot/share/opto/predicates.cpp

```diff
@@ -137,22 +137,19 @@ bool RuntimePredicate::is_predicate(Node* node, Deoptimization::DeoptReason deop
   return RegularPredicateWithUCT::is_predicate(node, deopt_reason);
 }
 
-// A Template Assertion Predicate has an If/RangeCheckNode and either an UCT or a halt node depending on where it
-// was created.
+// Template Assertion Predicates always have the dedicated OpaqueTemplateAssertionPredicate to identify them.
 bool TemplateAssertionPredicate::is_predicate(Node* node) {
   if (!may_be_assertion_predicate_if(node)) {
     return false;
   }
   IfNode* if_node = node->in(0)->as_If();
-  if (if_node->in(1)->is_Opaque4()) {
-    return RegularPredicateWithUCT::has_valid_uncommon_trap(node) || AssertionPredicateWithHalt::has_halt(node);
-  }
-  return false;
+  return if_node->in(1)->is_OpaqueTemplateAssertionPredicate();
```

Comment:
Detection is now easier. I plan to add more verification code that we also find the correct nodes on the failing path. But I'll wait with that until we have only Halt nodes for Template Assertion Predicates.